### PR TITLE
Add Rules API specification and integration to Observatorium API

### PR DESCRIPTION
**Pull request by jessicalins**
Ticket: https://issues.redhat.com/browse/MON-1965

This PR contains:
* The OpenAPI Spec for the Rules API
* The client and server code generated from the OpenAPI spec
* The integration of the Rules API into the main Observatorium API

An e2e test is missing, will be added in a separate PR.

TODO

- [x] Add alerting rule to spec

Signed-off-by: Jéssica Lins <jlins@redhat.com>

### CODE REVIEW

File: rules/spec.yaml
Comment (onprem) [reply rules/spec.yaml:13]: For now I went with the approach to only have a single file just for the sake of simplicity. Although the Rules backend is storing the files in object storage in a structure that'll allow us to move to multiple files per tenant without breaking anything.

Also, if we allow multiple files, I think it'd be great to have one endpoint that returns all rules merged in a single response. This will make the rule syncer's work much easier.

File: rules/server/server.go
Comment (bwplotka) [inline rules/server/server.go:15]: I wished someday we have gRPC defined APIs that by default expose  HTTP route too (like https://github.com/twitchtv/twirp) 😢

File: rules/spec.yaml
Comment (bwplotka) [reply rules/spec.yaml:13]: Sorry, I don't get that. The `$ref: '#/components/schemas/Rules'`  points that Rules are set of Groups, right? there is not file notion.

**Review by bwplotka — COMMENTED** [review]
Nice, some extra input (:

File: rules/spec.yaml
Comment (ianbillett) [reply rules/spec.yaml:13]: Ah ok - I understand this now. The identifier for the set of rules is the rule group name - tenants can choose to create and update any they like while keeping some level of semantic separation if they wish,

**Review by ianbillett — DISMISSED** [review]

File: rules/spec.yaml
Comment (jessicalins) [reply rules/spec.yaml]: changed it :)

File: rules/spec.yaml
Comment (jessicalins) [reply rules/spec.yaml:13]: I'm actually a bit confused - as I understand this still means one file per request, but using the `Rules` schema, that uses `RuleGroups`? is this what we want?
We can also test here: https://app.swaggerhub.com/apis/observatorium/rules-backend/1.0.0#/Rules

Comment (jessicalins) [reply rules/spec.yaml:13]: and here where we define yaml as content-type and the use of the `Rules` schema: https://github.com/observatorium/api/blob/44d2f6c58aebab68636f524b7ffdd5debcc304bd/rules/spec.yaml#L48-L53

File: rules/server/server.go
Comment (squat) [inline rules/server/server.go:33]: There's something funny going on here. One of the nice things of deepmap-oapi-gen is that it can generate two server interfaces for you, a basic HTTP server and also a typed, domain-aware server that unmarshals all of the requests into the correct Golang type you defined in your OpenAPI spec. Any idea why this isn't getting generated?

In any case, if we _did_ generate this, then the generated server would _have_ to be in the same package as the types (and this also the client), otherwise the server would't know what types to use.

Maybe this nice server isn't getting generated because you've put the server into a separate package?

I think it would be very worth dealing with the nicer server interface. This makes the generated server much more useful, as it automatically checks and unpacks all of the request parameters for you and puts them into the correct types, etc. This is also how grpc clients/servers are normally pacakged, so it's not an unfamiliar pattern

File: Makefile
Comment (squat) [reply Makefile:207]: As mentioned below, I think it would be worth generating all of the files into the same go package, as this should allow us to produce better server stubs. BTW, in this case you can collapse these three generation calls into one single command, e.g.
```sh
@oapi-codegen -generate types,client,chi-server -package rules -o rules/rules.go
```
This means also that the Makefile target would not need to be phony.

Comment (squat) [inline Makefile:207]: We might want to rename this as we are not generating a spec, but rather using a spec to generate code
```suggestion
generate-rules-code:
```

File: Makefile
Comment (squat) [inline Makefile:208]: We should definitely use Bingo to vendor and version oapi-gen as a tooling dependency of this project so that we can reproducibly regenerate code, so all devs are guaranteed to use the same oapi-gen version and CI can verify the generation
```suggestion
	$(OAPIGEN) -generate client -o rules/client.go -package rules rules/spec.yaml
```

File: rules/spec.yaml
Comment (bwplotka) [inline rules/spec.yaml:82]: Ok, I think the problem here is that we want to return alerting rules too. So we might want to have some union here.

Also ideally this is exactly the same as Prometheus Rules HTTP API, which has some specifics... ): 

We tried to improve this on Thanos by using protobuf and this did not pan-out really: https://thanos.io/tip/proposals-accepted/202107-protobuf-openapi-httpapi.md/

**Comment by jessicalins** [comment]
Holding this for now - I think we need to align:
- if the code generated from OpenAPI would work and conform with the Prometheus Rules endpoint API (probably not, as mentioned [here](https://github.com/observatorium/api/pull/178#discussion_r741091723))
- result from this ^ if we want to use OpenAPI at all

File: Makefile
Comment (jessicalins) [reply Makefile]: @squat I've added the oapi-gen using `bingo get github.com/deepmap/oapi-codegen/cmd/oapi-codegen@v1.9.0` and modified the Makefile too - but now I get this error on CI: 
```bash
/go/bin/jsonnet-v0.16.0 -m examples/manifests examples/main.jsonnet | xargs -I{} sh -c 'cat {} | /go/bin/gojsontoyaml-v0.0.0-20200602132005-3697ded27e8c > {}.yaml && rm -f {}' -- {
/go/bin/oapi-codegen-v1.9.0 -generate types,client,chi-server -package rules -o rules/rules.go rules/spec.yaml
bash: line 1: /go/bin/oapi-codegen-v1.9.0: No such file or directory
make: *** [Makefile:207: generate-rules-code] Error 127

Exited with code exit status 2
```
I think I missed some step but not sure yet..do you have any ideas?

File: Makefile
Comment (squat) [inline Makefile:206]: Yes, @jessicalins the missing piece here is that you need to add `$(OAPI_CODEGEN)` as a dependency of the generation target, otherwise the binary will not be guaranteed to exist

Comment (jessicalins) [reply Makefile]: thanks! added now

**Comment by jessicalins** [comment]
I've changed the spec/generated code to also support alerting rules (using https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/ as base) could you take another look?

**Review by bwplotka — APPROVED** [review]
This looks ok, but how we stand with compatibility with native Prometheus rule API at the end?

 What's the decision on our goal and status for this PR? 🤗 
 
 Otherwise LGTM

**Comment by squat** [comment]
> but how we stand with compatibility with native Prometheus rule API at the end?

During yesterday's community meeting we agreed that providing compete compatibility with the Prometheus rules API was out of scope for this proposal, which is geared towards providing an API for tenants to configure rules, not to read back what rules are firing etc. We agreed we should submit a new proposal to Observatorium to agree to implement compatibility with Prometheus's rules API.

File: Makefile
Comment (squat) [inline Makefile:206]: Let's make this a concrete target that produces a real file, rather than a phony target (or make it phony for convenience in addition to the real file). Let's also add this target as a dependency to the main API binary.
```suggestion
rules/rules.go: $(OAPI_CODEGEN)
```

Comment (jessicalins) [reply Makefile]: added (also had to make it .PHONY so the file would be correctly updated when running `make rules/rules.go` or `make generate`)

regarding add as a dependency from the main API bin, you mean here?
https://github.com/observatorium/api/blob/d1f53ca3b69122164d0db2097d224f5391b862f4/Makefile#L55

I think the wildcard would already catch the `rules/rules.go` file?

File: Makefile
Comment (squat) [inline Makefile:206]: This one should not be a `.PHONY` target, it produces a real file that can be tracked for updates
```suggestion
```

Comment (squat) [reply Makefile:206]: I think it should still be made an explicit dependency of the binary target. Consider what happens if you `rm rules/rules.go`. If you then do `make tmp/bin/observatorium-api`, the wildcard doesn't catch this file because the file does not exist on disk, so the rules.go file is note generated.

Comment (jessicalins) [reply Makefile:206]: if I don't make it .PHONY, when changing `spec.yaml` the `rules.go` is not regenerated/updated:
```bash
jessica@cupsofwonder:~/workspace/api|add-rules-api-spec⚡ ⇒  make rules/rules.go
make: 'rules/rules.go' is up to date.
```

same happens when using `make generate`..

good point with making the dependency explicit, will add

Comment (squat) [reply Makefile:206]: aha. Yes, this target should _not_ be marked as `.PHONY`. The missing piece here is that the target needs to depend on the `rules/spec.yaml`, this way whenever the spec is newer than the generated file, make will know that the rules.go file needs to be regenerated.

File: Makefile
Comment (squat) [inline Makefile:207]: From above's comment:
```suggestion
rules/rules.go: $(OAPI_CODEGEN) rules/spec.yaml
```

File: Makefile
Comment (squat) [inline Makefile:208]: This should likely be "simplified" as such so that the output file is always in sync with the target name
```suggestion
	$(OAPI_CODEGEN) -generate types,client,chi-server -package rules -o $@ rules/spec.yaml
```

File: Makefile
Comment (jessicalins) [reply Makefile:206]: nice, tested locally and worked :tada: thanks!

File: main.go
Comment (squat) [inline main.go:837]: The only last nit I have is that this is no longer _just_ recording rules, right? It can be alerting rules as well?

Comment (jessicalins) [reply main.go]: True! updated it

**Review by squat — APPROVED** [review]
LGTM good work @jessicalins :)

File: api/metrics/v1/http.go
Comment (onprem) [inline api/metrics/v1/http.go:291]: This route is same as the [Rules endpoint in Prom](https://prometheus.io/docs/prometheus/latest/querying/api/#rules) (also exposed by Thanos Querier) so it'll shadow it breaking anything (like Thanos Rules UI) that depends on it. I think we should change the `rulesRoute` to avoid confusion.

**Review by onprem — COMMENTED** [review]
Looking good to me, just one thing I'd like to get addressed before we merge. Great work @jessicalins!

File: api/metrics/v1/http.go
Comment (jessicalins) [reply api/metrics/v1/http.go:291]: very good point - I've changed it (now I think is in line with the [proposal](https://observatorium.io/docs/proposals/20201019-prometheus-rules-in-observatorium-api.md/) too)

Comment (jessicalins) [reply api/metrics/v1/http.go:291]: changed it to `api/v1/rulesfile` since either `api/v1/rules` and `/rules` would still create conflict

File: api/metrics/v1/http.go
Comment (squat) [inline api/metrics/v1/http.go:35]: naming is hard :see_no_evil: `/api/v1/rulesfile` sounds a bit funny to me though since when the user `put`s to the API, you're not really creating a file. hmm maybe `/api/v1/raw_rules`?

Comment (jessicalins) [reply api/metrics/v1/http.go:35]: it is indeed :sweat_smile: we thought `rulesfile` because the content has the same format as the Prometheus rules file - but I agree that gets confusing.. I'm fine with `raw_rules` too @onprem what do you think?

Comment (onprem) [reply api/metrics/v1/http.go:35]: Yep that sounds good. Another one I can think of from what Lucas suggested is `/api/v1/rules/raw`, which just feels a bit more natural to me. WDYT @squat @jessicalins ?

Comment (jessicalins) [reply api/metrics/v1/http.go:35]: ^ I like it :) aligns better with the other rules endpoints too

Comment (squat) [reply api/metrics/v1/http.go:35]: I second this

File: api/metrics/v1/rules.go
Comment (onprem) [inline api/metrics/v1/rules.go:24]: Maybe it makes sense to propagate the HTTP status code we got from the rules backend instead of `500`. I am thinking about the cases when it fails some validation checks on Rules Backend side and it returns a `400 Bad Request`.

File: api/metrics/v1/rules.go
Comment (onprem) [inline api/metrics/v1/rules.go:49]: ditto.

**Review by onprem — APPROVED** [review]
Looking great to me! Just a non-blocking suggestion we can iterate on future PRs.

**Review by squat — APPROVED** [review]

**Comment by squat** [comment]
Great work @jessicalins :tada: :taco: Let's get this merged

**Review by onprem — APPROVED** [review]
Great work! :100:

**Pull request merged**
